In [ ]:
import pandas as pd

## 1.&nbsp; Import a csv file to DataFrame

In [ ]:
url = "https://drive.google.com/file/d/1FYhN_2AzTBFuWcfHaRuKcuCE6CWXsWtG/view?usp=sharing" # orderlines.csv
path = "https://drive.google.com/uc?export=download&id="+url.split("/")[-2]
df = pd.read_csv(path)

## 2.&nbsp; DataFrame dimensions

With the [DataFrame.shape](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.shape.html) attribute we can calculate the dimensions (number of rows and columns) of the DataFrame.

In [ ]:
df.shape

(293983, 7)

As a result we obtain a [tuple](https://www.w3schools.com/python/python_tuples.asp) where the first element is the number of rows, which in our case is 293983, while the second element is the number of columns, which in our example was 7.

In [ ]:
nrows = df.shape[0]
ncols = df.shape[1]
print("The number of rows is", nrows)
print("The number of columns is", ncols)

The number of rows is 293983
The number of columns is 7


[DataFrame.size](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.size.html) returns the total number of values that the DataFrame has (the number of rows multiplied by the number of columns):

In [ ]:
df.size

2057881

We can check if the `.size` and `.shape` agree

In [ ]:
df.shape[0] * df.shape[1] == df.size

True

With the [DataFrame.ndim](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.ndim.html) attribute we calculate the number of dimensions that the DataFrame has. This will always be 2, as it consists of rows and columns.

In [ ]:
df.ndim

2

## 3.&nbsp; DataFrames exploration

The [DataFrame.head()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.head.html) and [DataFrame.tail()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.tail.html) methods are used to display the first or last rows of the DataFrame. Looking at the raw data is a great way to get a grasp of what's in there. By default, 5 rows will be shown, but you can change that:

In [ ]:
df.head()

id  id_order  product_id  product_quantity      sku unit_price  \
0  1119109    299539           0                 1  OTT0133      18.99   
1  1119110    299540           0                 1  LGE0043     399.00   
2  1119111    299541           0                 1  PAR0071     474.05   
3  1119112    299542           0                 1  WDT0315      68.39   
4  1119113    299543           0                 1  JBL0104      23.74   

                  date  
0  2017-01-01 00:07:19  
1  2017-01-01 00:19:45  
2  2017-01-01 00:20:57  
3  2017-01-01 00:51:40  
4  2017-01-01 01:06:38

In [ ]:
df.head(9)

id  id_order  product_id  product_quantity      sku unit_price  \
0  1119109    299539           0                 1  OTT0133      18.99   
1  1119110    299540           0                 1  LGE0043     399.00   
2  1119111    299541           0                 1  PAR0071     474.05   
3  1119112    299542           0                 1  WDT0315      68.39   
4  1119113    299543           0                 1  JBL0104      23.74   
5  1119114    295310           0                10  WDT0249     231.79   
6  1119115    299544           0                 1  APP1582   1.137.99   
7  1119116    299545           0                 1  OWC0100      47.49   
8  1119119    299546           0                 1  IOT0014      18.99   

                  date  
0  2017-01-01 00:07:19  
1  2017-01-01 00:19:45  
2  2017-01-01 00:20:57  
3  2017-01-01 00:51:40  
4  2017-01-01 01:06:38  
5  2017-01-01 01:14:27  
6  2017-01-01 01:17:21  
7  2017-01-01 01:46:16  
8  2017-01-01 01:50:34

In [ ]:
df.tail()

id  id_order  product_id  product_quantity      sku unit_price  \
293978  1650199    527398           0                 1  JBL0122      42.99   
293979  1650200    527399           0                 1  PAC0653     141.58   
293980  1650201    527400           0                 2  APP0698       9.99   
293981  1650202    527388           0                 1  BEZ0204      19.99   
293982  1650203    527401           0                 1  APP0927      13.99   

                       date  
293978  2018-03-14 13:57:25  
293979  2018-03-14 13:57:34  
293980  2018-03-14 13:57:41  
293981  2018-03-14 13:58:01  
293982  2018-03-14 13:58:36

The methods [DataFrame.info()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.info.html), [DataFrame.describe()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.describe.html) and [DataFrame.nunique()](https://pandas.pydata.org/docs/reference/api/pandas.core.groupby.DataFrameGroupBy.nunique.html) give a general overview of what's in the DataFrame:

`.info()` tells us about how pandas sees the data - how it is stored, whether there's any missing values, and how many columns and rows we have.

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293983 entries, 0 to 293982
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   id                293983 non-null  int64 
 1   id_order          293983 non-null  int64 
 2   product_id        293983 non-null  int64 
 3   product_quantity  293983 non-null  int64 
 4   sku               293983 non-null  object
 5   unit_price        293983 non-null  object
 6   date              293983 non-null  object
dtypes: int64(4), object(3)
memory usage: 15.7+ MB


`.describe()` gives us an overview of the [descriptive statistics](https://www.scribbr.com/statistics/descriptive-statistics/) for the numerical columns of our DataFrame.

In [ ]:
df.describe()

id       id_order  product_id  product_quantity
count  2.939830e+05  293983.000000    293983.0     293983.000000
mean   1.397918e+06  419999.116544         0.0          1.121126
std    1.530096e+05   66344.486479         0.0          3.396569
min    1.119109e+06  241319.000000         0.0          1.000000
25%    1.262542e+06  362258.500000         0.0          1.000000
50%    1.406940e+06  425956.000000         0.0          1.000000
75%    1.531322e+06  478657.000000         0.0          1.000000
max    1.650203e+06  527401.000000         0.0        999.000000

In [ ]:
df.nunique()

id                  293983
id_order            204855
product_id               1
product_quantity        67
sku                   7951
unit_price           11329
date                251631
dtype: int64

If we wish, we can also view the unique values counted by `series.nunique()`. The [.unique()](https://pandas.pydata.org/docs/reference/api/pandas.Series.unique.html) method returns the unique values from a column as a numpy array, which can be indexed with `[]`:

In [ ]:
df['product_id'].unique()

array([0])

In [ ]:
df["sku"].unique()[:10]

array(['OTT0133', 'LGE0043', 'PAR0071', 'WDT0315', 'JBL0104', 'WDT0249',
       'APP1582', 'OWC0100', 'IOT0014', 'APP0700'], dtype=object)

The [DataFrame.isna()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.isna.html) method returns a boolean for each value: `True` if that value is "missing" (which is represented as `NaN` in numpy and pandas) and `False` if the value is not missing:

In [ ]:
df.isna()

id  id_order  product_id  product_quantity    sku  unit_price  \
0       False     False       False             False  False       False   
1       False     False       False             False  False       False   
2       False     False       False             False  False       False   
3       False     False       False             False  False       False   
4       False     False       False             False  False       False   
...       ...       ...         ...               ...    ...         ...   
293978  False     False       False             False  False       False   
293979  False     False       False             False  False       False   
293980  False     False       False             False  False       False   
293981  False     False       False             False  False       False   
293982  False     False       False             False  False       False   

         date  
0       False  
1       False  
2       False  
3       False  
4       False  
...       ...  
293978  False  
293979  False  
293980  False  
293981  False  
293982  False  

[293983 rows x 7 columns]

We can then use [DataFrame.sum()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.sum.html) to add up all these booleans for each column, and count how many missing values are there in the DataFrame, since `True` is interpreted as `1` and `False` as `0`:

In [ ]:
df.isna().sum()

id                  0
id_order            0
product_id          0
product_quantity    0
sku                 0
unit_price          0
date                0
dtype: int64

[DataFrame.duplicated()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.duplicated.html) also returns a boolean output, but in this case just one value per row: `True` if that row is duplicated and `False` if it's not. Again, using `sum()` allows us to count how many `True` values (i.e. duplicated rows) are there in total:

In [ ]:
df.duplicated().sum()

0

[DataFrame.nlargest(n, columns)](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.nlargest.html) will return the top `n` rows with the largest value for whatever column we specify in `columns`. Below, we see the rows with the largest product quantity values:

In [ ]:
df.nlargest(5, "product_quantity")

id  id_order  product_id  product_quantity      sku unit_price  \
53860  1228150    346221           0               999  APP1190      55.99   
68712  1254032    358747           0               999  SEV0028      19.99   
57796  1234924    349475           0               800  KIN0137       7.49   
57306  1234111    349133           0               555  APP0665      70.99   
40813  1204788    335057           0               201  THU0029      80.99   

                      date  
53860  2017-04-14 21:50:52  
68712  2017-05-24 14:51:58  
57796  2017-04-25 09:59:00  
57306  2017-04-24 10:20:13  
40813  2017-03-14 15:25:53

[DataFrame.nsmallest()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.nsmallest.html) does the same, for the smallest values:

In [ ]:
df.nsmallest(5, "product_quantity")

id  id_order  product_id  product_quantity      sku unit_price  \
0  1119109    299539           0                 1  OTT0133      18.99   
1  1119110    299540           0                 1  LGE0043     399.00   
2  1119111    299541           0                 1  PAR0071     474.05   
3  1119112    299542           0                 1  WDT0315      68.39   
4  1119113    299543           0                 1  JBL0104      23.74   

                  date  
0  2017-01-01 00:07:19  
1  2017-01-01 00:19:45  
2  2017-01-01 00:20:57  
3  2017-01-01 00:51:40  
4  2017-01-01 01:06:38

So far, we have seen two ways to explore DataFrames:

* **Attributes:** `.shape`, `.size` and `.ndim`, and others. They are written without parentheses and give you raw "metadata" about the DataFrame you are calling them on.
* **Methods:** `.head()`, `.describe()` and `.isna()`, and others. They are written with parentheses and perform some sort of calculation, transformation or aggregation. A method is like a function that is tied to a specific object type.

DataFrames have a lot of attributes and methods and may not be obvious whether something belongs to one type or the other. Whenever in doubt, check [the documentation](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html?highlight=DataFrame#pandas.DataFrame). If you scroll down past the list of examples you will find two sections listing all the attributes and methods.

## 4.&nbsp; Select Columns

Given a DataFrame, we can select a particular column in several ways:

* Indicating the name of the column between square brackets, `[]`
* With the `.loc[]` attribute (by name or tag)
* With the `.iloc[]` attribute (by position)

Plain square brackets `[]` are used to just view a column if you know its name and **don't want to modify it**:

In [ ]:
df["id_order"]

0         299539
1         299540
2         299541
3         299542
4         299543
           ...  
293978    527398
293979    527399
293980    527400
293981    527388
293982    527401
Name: id_order, Length: 293983, dtype: int64

`.loc[]` takes two arguments: `[rows, columns]`. Passing `:` to the rows argument means "grabbing all the rows", which allows you to select a whole column, if you know its name. This method is more flexible as you will see in the future, and allows you to modify the data.

We recommend using `loc[]` as the primary option for selecting data.

In [ ]:
df.loc[:, "id_order"]

0         299539
1         299540
2         299541
3         299542
4         299543
           ...  
293978    527398
293979    527399
293980    527400
293981    527388
293982    527401
Name: id_order, Length: 293983, dtype: int64

`.iloc[]` works similarly to `.loc[]`, but it only accepts integers, which represent the positions of the rows and columns:

In [ ]:
df.iloc[:, 1]

0         299539
1         299540
2         299541
3         299542
4         299543
           ...  
293978    527398
293979    527399
293980    527400
293981    527388
293982    527401
Name: id_order, Length: 293983, dtype: int64

### 4.1.&nbsp; Select multiple columns

If we want to select more than one column, we can do it with all the options listed above, with slight modifications in some cases:

> **Note:** we pass a list inside of the `[]`, a list is also represented by `[]`, hence why we have 2 sets of square brackets

In [ ]:
df[["id_order","sku"]]

id_order      sku
0         299539  OTT0133
1         299540  LGE0043
2         299541  PAR0071
3         299542  WDT0315
4         299543  JBL0104
...          ...      ...
293978    527398  JBL0122
293979    527399  PAC0653
293980    527400  APP0698
293981    527388  BEZ0204
293982    527401  APP0927

[293983 rows x 2 columns]

In [ ]:
df.loc[:, ["id_order", "sku"]]

id_order      sku
0         299539  OTT0133
1         299540  LGE0043
2         299541  PAR0071
3         299542  WDT0315
4         299543  JBL0104
...          ...      ...
293978    527398  JBL0122
293979    527399  PAC0653
293980    527400  APP0698
293981    527388  BEZ0204
293982    527401  APP0927

[293983 rows x 2 columns]

With `.loc[]` and `:` you can select all columns between two columns you specify.

In [ ]:
df.loc[:, "id_order":"sku"]

id_order  product_id  product_quantity      sku
0         299539           0                 1  OTT0133
1         299540           0                 1  LGE0043
2         299541           0                 1  PAR0071
3         299542           0                 1  WDT0315
4         299543           0                 1  JBL0104
...          ...         ...               ...      ...
293978    527398           0                 1  JBL0122
293979    527399           0                 1  PAC0653
293980    527400           0                 2  APP0698
293981    527388           0                 1  BEZ0204
293982    527401           0                 1  APP0927

[293983 rows x 4 columns]

In [ ]:
df.iloc[:, [1, 4]]

id_order      sku
0         299539  OTT0133
1         299540  LGE0043
2         299541  PAR0071
3         299542  WDT0315
4         299543  JBL0104
...          ...      ...
293978    527398  JBL0122
293979    527399  PAC0653
293980    527400  APP0698
293981    527388  BEZ0204
293982    527401  APP0927

[293983 rows x 2 columns]

In [ ]:
df.iloc[:, 1:5]

id_order  product_id  product_quantity      sku
0         299539           0                 1  OTT0133
1         299540           0                 1  LGE0043
2         299541           0                 1  PAR0071
3         299542           0                 1  WDT0315
4         299543           0                 1  JBL0104
...          ...         ...               ...      ...
293978    527398           0                 1  JBL0122
293979    527399           0                 1  PAC0653
293980    527400           0                 2  APP0698
293981    527388           0                 1  BEZ0204
293982    527401           0                 1  APP0927

[293983 rows x 4 columns]

## 5.&nbsp; Select Rows

Selecting rows is easy if you know how to select columns. You have two options:

* With `.loc[]` (by name or tag)
* With `.iloc[]` (by position)

Selecting a single row returns a pandas Series (a 1-dimensional object):

In [ ]:
df.loc[0, :]

id                              1119109
id_order                         299539
product_id                            0
product_quantity                      1
sku                             OTT0133
unit_price                        18.99
date                2017-01-01 00:07:19
Name: 0, dtype: object

With `.loc[]`, rows are selected by their index name:

In [ ]:
df.loc[0:3, :]

id  id_order  product_id  product_quantity      sku unit_price  \
0  1119109    299539           0                 1  OTT0133      18.99   
1  1119110    299540           0                 1  LGE0043     399.00   
2  1119111    299541           0                 1  PAR0071     474.05   
3  1119112    299542           0                 1  WDT0315      68.39   

                  date  
0  2017-01-01 00:07:19  
1  2017-01-01 00:19:45  
2  2017-01-01 00:20:57  
3  2017-01-01 00:51:40

If we change the index and set it to the `id` column, now the first rows can not be selected the same way.

In [ ]:
df = df.set_index("id")

Look, the same code as above no longer works. The index has been changed to the `id` column, if we want to use `.loc` we now need to use the names/numbers from `id`

In [ ]:
df.loc[0:3, :]

Empty DataFrame
Columns: [id_order, product_id, product_quantity, sku, unit_price, date]
Index: []

In [ ]:
df.loc[1119109:1119112, :]

id_order  product_id  product_quantity      sku unit_price  \
id                                                                    
1119109    299539           0                 1  OTT0133      18.99   
1119110    299540           0                 1  LGE0043     399.00   
1119111    299541           0                 1  PAR0071     474.05   
1119112    299542           0                 1  WDT0315      68.39   

                        date  
id                            
1119109  2017-01-01 00:07:19  
1119110  2017-01-01 00:19:45  
1119111  2017-01-01 00:20:57  
1119112  2017-01-01 00:51:40

You can always tell the index column as it will be in bold

With the `.iloc[]` method we don't need to know the row names to select rows at a certain position. Every row and column is instead nominally indexed, starting from the number 0:

In [ ]:
df.iloc[0:4]

id_order  product_id  product_quantity      sku unit_price  \
id                                                                    
1119109    299539           0                 1  OTT0133      18.99   
1119110    299540           0                 1  LGE0043     399.00   
1119111    299541           0                 1  PAR0071     474.05   
1119112    299542           0                 1  WDT0315      68.39   

                        date  
id                            
1119109  2017-01-01 00:07:19  
1119110  2017-01-01 00:19:45  
1119111  2017-01-01 00:20:57  
1119112  2017-01-01 00:51:40

We can select the last observation with the method `.iloc[]` by using negative numbers

In [ ]:
df.tail()

id_order  product_id  product_quantity      sku unit_price  \
id                                                                    
1650199    527398           0                 1  JBL0122      42.99   
1650200    527399           0                 1  PAC0653     141.58   
1650201    527400           0                 2  APP0698       9.99   
1650202    527388           0                 1  BEZ0204      19.99   
1650203    527401           0                 1  APP0927      13.99   

                        date  
id                            
1650199  2018-03-14 13:57:25  
1650200  2018-03-14 13:57:34  
1650201  2018-03-14 13:57:41  
1650202  2018-03-14 13:58:01  
1650203  2018-03-14 13:58:36

In [ ]:
df.iloc[[-2],:]

id_order  product_id  product_quantity      sku unit_price  \
id                                                                    
1650202    527388           0                 1  BEZ0204      19.99   

                        date  
id                            
1650202  2018-03-14 13:58:01

We can, whenever necessary, reset the index to a set of numbers starting from 0. This would also make the `id` column a normal column once again

In [ ]:
# df.reset_index(inplace=True)
df = df.reset_index()

Indexing can get tricky sometimes, it's ok to take some time to get used to the methods we presented, and it's ok to have some trouble selecting the rows and columns you need. For an exhaustive guide on Pandas indexing, check out this link: https://pandas.pydata.org/docs/user_guide/indexing.html#indexing

## 6.&nbsp; Drop Columns

The `.drop()` method allows us to delete the rows or columns.

> **Note:** Again, if we want to directly apply the changes to the original DataFrame, we need to indicate `inplace = True`. Otherwise, we are getting as an output just a "view" of how the DataFrame looks like after the drop, but the original DataFrame remains unchanged.

> **Note:** `axis=1` means we want to drop a column, not a row. For rows we would use `axis=0`

In [ ]:
df.drop(["product_id"],axis=1)

id  id_order  product_quantity      sku unit_price  \
0       1119109    299539                 1  OTT0133      18.99   
1       1119110    299540                 1  LGE0043     399.00   
2       1119111    299541                 1  PAR0071     474.05   
3       1119112    299542                 1  WDT0315      68.39   
4       1119113    299543                 1  JBL0104      23.74   
...         ...       ...               ...      ...        ...   
293978  1650199    527398                 1  JBL0122      42.99   
293979  1650200    527399                 1  PAC0653     141.58   
293980  1650201    527400                 2  APP0698       9.99   
293981  1650202    527388                 1  BEZ0204      19.99   
293982  1650203    527401                 1  APP0927      13.99   

                       date  
0       2017-01-01 00:07:19  
1       2017-01-01 00:19:45  
2       2017-01-01 00:20:57  
3       2017-01-01 00:51:40  
4       2017-01-01 01:06:38  
...                     ...  
293978  2018-03-14 13:57:25  
293979  2018-03-14 13:57:34  
293980  2018-03-14 13:57:41  
293981  2018-03-14 13:58:01  
293982  2018-03-14 13:58:36  

[293983 rows x 6 columns]

In [ ]:
df.drop(columns='product_id')

id  id_order  product_quantity      sku unit_price  \
0       1119109    299539                 1  OTT0133      18.99   
1       1119110    299540                 1  LGE0043     399.00   
2       1119111    299541                 1  PAR0071     474.05   
3       1119112    299542                 1  WDT0315      68.39   
4       1119113    299543                 1  JBL0104      23.74   
...         ...       ...               ...      ...        ...   
293978  1650199    527398                 1  JBL0122      42.99   
293979  1650200    527399                 1  PAC0653     141.58   
293980  1650201    527400                 2  APP0698       9.99   
293981  1650202    527388                 1  BEZ0204      19.99   
293982  1650203    527401                 1  APP0927      13.99   

                       date  
0       2017-01-01 00:07:19  
1       2017-01-01 00:19:45  
2       2017-01-01 00:20:57  
3       2017-01-01 00:51:40  
4       2017-01-01 01:06:38  
...                     ...  
293978  2018-03-14 13:57:25  
293979  2018-03-14 13:57:34  
293980  2018-03-14 13:57:41  
293981  2018-03-14 13:58:01  
293982  2018-03-14 13:58:36  

[293983 rows x 6 columns]

See how our output above shows no column `product_id`. Then look at the output below, `product_id` is magically back. We didn't use `inplace=True` above, which means that our output displayed what we asked, but didn't alter the original DataFrame. Be careful of this, Pandas only changes data if you explicitly tell it to do so.

In [ ]:
df

id  id_order  product_id  product_quantity      sku unit_price  \
0       1119109    299539           0                 1  OTT0133      18.99   
1       1119110    299540           0                 1  LGE0043     399.00   
2       1119111    299541           0                 1  PAR0071     474.05   
3       1119112    299542           0                 1  WDT0315      68.39   
4       1119113    299543           0                 1  JBL0104      23.74   
...         ...       ...         ...               ...      ...        ...   
293978  1650199    527398           0                 1  JBL0122      42.99   
293979  1650200    527399           0                 1  PAC0653     141.58   
293980  1650201    527400           0                 2  APP0698       9.99   
293981  1650202    527388           0                 1  BEZ0204      19.99   
293982  1650203    527401           0                 1  APP0927      13.99   

                       date  
0       2017-01-01 00:07:19  
1       2017-01-01 00:19:45  
2       2017-01-01 00:20:57  
3       2017-01-01 00:51:40  
4       2017-01-01 01:06:38  
...                     ...  
293978  2018-03-14 13:57:25  
293979  2018-03-14 13:57:34  
293980  2018-03-14 13:57:41  
293981  2018-03-14 13:58:01  
293982  2018-03-14 13:58:36  

[293983 rows x 7 columns]

## 7.&nbsp; Filter rows based on conditions

Using `.loc` to filter for rows that have a `product_quantity greater than 100`

In [ ]:
df.loc[df["product_quantity"] > 100, :]

id  id_order  product_id  product_quantity      sku unit_price  \
27779   1180010    323959           0               126  ADN0039      34.99   
40813   1204788    335057           0               201  THU0029      80.99   
53860   1228150    346221           0               999  APP1190      55.99   
57306   1234111    349133           0               555  APP0665      70.99   
57796   1234924    349475           0               800  KIN0137       7.49   
68712   1254032    358747           0               999  SEV0028      19.99   
136675  1388261    417536           0               200  TRK0009      29.99   
204637  1500715    464858           0               192  APP1662     519.00   
246048  1574262    496172           0               164  EVU0013      19.99   
285492  1637611    522075           0               125  XDO0047      25.99   

                       date  
27779   2017-02-14 10:21:12  
40813   2017-03-14 15:25:53  
53860   2017-04-14 21:50:52  
57306   2017-04-24 10:20:13  
57796   2017-04-25 09:59:00  
68712   2017-05-24 14:51:58  
136675  2017-10-25 15:02:39  
204637  2017-12-17 15:53:04  
246048  2018-01-22 16:14:42  
285492  2018-03-06 10:07:54

In [ ]:
mask_quantity = df["product_quantity"] > 100
df.loc[mask_quantity,:]

id  id_order  product_id  product_quantity      sku unit_price  \
27779   1180010    323959           0               126  ADN0039      34.99   
40813   1204788    335057           0               201  THU0029      80.99   
53860   1228150    346221           0               999  APP1190      55.99   
57306   1234111    349133           0               555  APP0665      70.99   
57796   1234924    349475           0               800  KIN0137       7.49   
68712   1254032    358747           0               999  SEV0028      19.99   
136675  1388261    417536           0               200  TRK0009      29.99   
204637  1500715    464858           0               192  APP1662     519.00   
246048  1574262    496172           0               164  EVU0013      19.99   
285492  1637611    522075           0               125  XDO0047      25.99   

                       date  
27779   2017-02-14 10:21:12  
40813   2017-03-14 15:25:53  
53860   2017-04-14 21:50:52  
57306   2017-04-24 10:20:13  
57796   2017-04-25 09:59:00  
68712   2017-05-24 14:51:58  
136675  2017-10-25 15:02:39  
204637  2017-12-17 15:53:04  
246048  2018-01-22 16:14:42  
285492  2018-03-06 10:07:54

The `.query()` method can be useful for this purpose. Some of you may prefer it as it resembles SQL syntax. You can use any **Python Comparison Operators** you want inside the query method (find more information on this [link](https://www.w3schools.com/python/python_operators.asp)).
> **Note:** `.query()` only works when the named column does not contain any white space and string values are contained within a separate set of quotes (the string value may contain white space). For example, if you have a column `a b` and you wish to query it `df.query("a b == 5")`, this would throw up an error. For column names you should always be using an underscore `_` not a space.

In [ ]:
df.query("product_quantity > 100")

id  id_order  product_id  product_quantity      sku unit_price  \
27779   1180010    323959           0               126  ADN0039      34.99   
40813   1204788    335057           0               201  THU0029      80.99   
53860   1228150    346221           0               999  APP1190      55.99   
57306   1234111    349133           0               555  APP0665      70.99   
57796   1234924    349475           0               800  KIN0137       7.49   
68712   1254032    358747           0               999  SEV0028      19.99   
136675  1388261    417536           0               200  TRK0009      29.99   
204637  1500715    464858           0               192  APP1662     519.00   
246048  1574262    496172           0               164  EVU0013      19.99   
285492  1637611    522075           0               125  XDO0047      25.99   

                       date  
27779   2017-02-14 10:21:12  
40813   2017-03-14 15:25:53  
53860   2017-04-14 21:50:52  
57306   2017-04-24 10:20:13  
57796   2017-04-25 09:59:00  
68712   2017-05-24 14:51:58  
136675  2017-10-25 15:02:39  
204637  2017-12-17 15:53:04  
246048  2018-01-22 16:14:42  
285492  2018-03-06 10:07:54

The `isin()` method is very useful to find rows that match any of the values you have **in a list**. For example, here we are searching for rows where its `sku` matches any of the 2 sku's we listed:

In [ ]:
# find out a column that contains a list
df["sku"].isin(["JBL0104", "ADN0039"])

0         False
1         False
2         False
3         False
4          True
          ...  
293978    False
293979    False
293980    False
293981    False
293982    False
Name: sku, Length: 293983, dtype: bool

This expression can be used inside of `[]` or `.loc[]` to filter the rows that have a `True` value. This is called "boolean indexing" and it is really useful:

In [ ]:
df.loc[df["sku"].isin(["JBL0104", "ADN0039"]), :]

id  id_order  product_id  product_quantity      sku unit_price  \
4       1119113    299543           0                 1  JBL0104      23.74   
716     1120463    300212           0                 2  JBL0104      26.99   
774     1120607    300272           0                 1  JBL0104      24.99   
866     1120758    300349           0                 1  JBL0104      24.99   
1411    1122014    300851           0                 1  JBL0104      24.99   
...         ...       ...         ...               ...      ...        ...   
258180  1593360    503247           0                 1  JBL0104      23.99   
263790  1601885    506553           0                 1  JBL0104      24.99   
266635  1606326    508345           0                 2  JBL0104      24.99   
268446  1608949    509377           0                 1  JBL0104      24.99   
279801  1628218    517916           0                 1  JBL0104      27.99   

                       date  
4       2017-01-01 01:06:38  
716     2017-01-02 01:44:05  
774     2017-01-02 08:57:07  
866     2017-01-02 10:41:29  
1411    2017-01-02 18:50:13  
...                     ...  
258180  2018-01-31 20:55:50  
263790  2018-02-05 22:41:40  
266635  2018-02-08 19:20:21  
268446  2018-02-10 18:02:37  
279801  2018-02-26 20:51:30  

[65 rows x 7 columns]

Pandas compresses large outputs in Colab/Jupyter Notebooks. If you want to see more rows, you can change the options:
> **Note:** after running this cell, run the code above again and you will see all the rows

In [ ]:
# pd.reset_option("^display")

In [ ]:
pd.options.display.max_rows = 100

For a complete list of all settings and options that can be tweaked, check out this: https://pandas.pydata.org/pandas-docs/stable/user_guide/options.html

In [ ]:
df.loc[df["sku"].isin(["JBL0104", "ADN0039"]), :]

id  id_order  product_id  product_quantity      sku unit_price  \
4       1119113    299543           0                 1  JBL0104      23.74   
716     1120463    300212           0                 2  JBL0104      26.99   
774     1120607    300272           0                 1  JBL0104      24.99   
866     1120758    300349           0                 1  JBL0104      24.99   
1411    1122014    300851           0                 1  JBL0104      24.99   
1451    1122093    300887           0                 1  JBL0104      24.99   
3612    1130604    302848           0                 1  JBL0104      24.99   
4370    1132185    303586           0                 1  JBL0104      22.31   
7028    1137895    305980           0                 1  ADN0039      34.99   
8450    1140943    307256           0                 1  JBL0104      26.99   
14439   1153300    312363           0                 1  JBL0104      26.99   
27779   1180010    323959           0               126  ADN0039      34.99   
40038   1203343    334417           0                 1  ADN0039      34.99   
53150   1226898    345622           0                 1  JBL0104      26.99   
56740   1233192    348682           0                 1  JBL0104      26.99   
67137   1251234    357411           0                 1  ADN0039      34.99   
69235   1254967    359135           0                 1  ADN0039      34.99   
98730   1307785    384545           0                 1  JBL0104      26.99   
105968  1320675    390750           0                 1  JBL0104      26.99   
112369  1344362    396145           0                 1  JBL0104      26.99   
113454  1346279    397061           0                 1  ADN0039      34.99   
154201  1423772    432828           0                 1  JBL0104      22.99   
165640  1443880    442197           0                 1  JBL0104      22.99   
165688  1443963    442228           0                 1  JBL0104      22.99   
171548  1453081    446003           0                 1  JBL0104      22.99   
172402  1454757    446688           0                 1  JBL0104      22.99   
175048  1459388    448846           0                 1  ADN0039      29.74   
192427  1482783    457514           0                 1  JBL0104      24.99   
195671  1486865    459056           0                 1  JBL0104      24.99   
196563  1488065    459522           0                 1  JBL0104      24.99   
203901  1499365    464210           0                 1  JBL0104      24.99   
205833  1503035    465885           0                 1  JBL0104      24.99   
206138  1503671    466170           0                 1  JBL0104      24.99   
206312  1503978    466344           0                 1  JBL0104      24.99   
206905  1505114    466844           0                 1  JBL0104      24.99   
207094  1505489    467005           0                 1  JBL0104      24.99   
207544  1506336    467370           0                 1  JBL0104      24.99   
211840  1514696    471338           0                 1  JBL0104      23.74   
212820  1516518    472215           0                 1  JBL0104      23.74   
213153  1517116    472506           0                 1  JBL0104      23.74   
213262  1517297    472597           0                 1  JBL0104      23.74   
213808  1518224    473037           0                 1  JBL0104      23.74   
213841  1518286    473074           0                 1  JBL0104      24.99   
214057  1518672    473261           0                 1  JBL0104      23.74   
214559  1519654    473732           0                 1  JBL0104      23.74   
216307  1522981    475353           0                 1  JBL0104      23.74   
217798  1525891    476696           0                 1  JBL0104      23.74   
218027  1526318    476911           0                 1  JBL0104      23.74   
219284  1528937    478069           0                 1  JBL0104      23.74   
220630  1531600    479235           0                 1  JBL010

#### 7.1.&nbsp; Modifying a DataFrame & the `.copy()` method

Let's take a small sample of data from our DataFrame:

In [ ]:
sample = df.iloc[:3,:]
sample

id  id_order  product_id  product_quantity      sku unit_price  \
0  1119109    299539           0                 1  OTT0133      18.99   
1  1119110    299540           0                 1  LGE0043     399.00   
2  1119111    299541           0                 1  PAR0071     474.05   

                  date  
0  2017-01-01 00:07:19  
1  2017-01-01 00:19:45  
2  2017-01-01 00:20:57

Now we pick a single cell from the sample we took and we assign a new value to it.

In [ ]:
sample.iloc[0,4] = "NEW VALUE HERE"

We can see the new value on the `sample` we took:

In [ ]:
sample

id  id_order  product_id  product_quantity             sku unit_price  \
0  1119109    299539           0                 1  NEW VALUE HERE      18.99   
1  1119110    299540           0                 1         LGE0043     399.00   
2  1119111    299541           0                 1         PAR0071     474.05   

                  date  
0  2017-01-01 00:07:19  
1  2017-01-01 00:19:45  
2  2017-01-01 00:20:57

..and, maybe to your surprise, we can see that the new value is also present on the original `df`!

In [ ]:
df.head()

id  id_order  product_id  product_quantity             sku unit_price  \
0  1119109    299539           0                 1  NEW VALUE HERE      18.99   
1  1119110    299540           0                 1         LGE0043     399.00   
2  1119111    299541           0                 1         PAR0071     474.05   
3  1119112    299542           0                 1         WDT0315      68.39   
4  1119113    299543           0                 1         JBL0104      23.74   

                  date  
0  2017-01-01 00:07:19  
1  2017-01-01 00:19:45  
2  2017-01-01 00:20:57  
3  2017-01-01 00:51:40  
4  2017-01-01 01:06:38

When you take a sample of data using `.loc[]` or `iloc[]` and assign it to a new object, the new object is just a "tag" pointing to the very same data as the original DataFrame points to. We can avoid this using the method `.copy()`

In [ ]:
url = "https://drive.google.com/file/d/14Y7g5ITyf6LMyPoKc9wr010V9StaCUux/view?usp=sharing" # orderlines.csv
path = "https://drive.google.com/uc?export=download&id="+url.split("/")[-2]
df = pd.read_csv(path)

In [ ]:
sample = df.iloc[:3,:].copy()
sample.iloc[0,4] = "NEW VALUE HERE"
sample

id  id_order  product_id  product_quantity             sku unit_price  \
0  1119109    299539           0                 1  NEW VALUE HERE      18.99   
1  1119110    299540           0                 1         LGE0043     399.00   
2  1119111    299541           0                 1         PAR0071     474.05   

                  date  
0  2017-01-01 00:07:19  
1  2017-01-01 00:19:45  
2  2017-01-01 00:20:57

In [ ]:
df.head(3)

id  id_order  product_id  product_quantity      sku unit_price  \
0  1119109    299539           0                 1  OTT0133      18.99   
1  1119110    299540           0                 1  LGE0043     399.00   
2  1119111    299541           0                 1  PAR0071     474.05   

                  date  
0  2017-01-01 00:07:19  
1  2017-01-01 00:19:45  
2  2017-01-01 00:20:57

As you can see, now it has not been modified.

In [ ]:
df.loc[df['id_order'].isna(),:]

Empty DataFrame
Columns: [id, id_order, product_id, product_quantity, sku, unit_price, date]
Index: []

# CHALLENGES

## Challenge 1:
How many different unit prices does the product with the sku JBL0104 have? Combine a pandas filter method with the method `.nunique()`.

In [ ]:
df.loc[df['sku']=='JBL0104',['unit_price']].nunique()

unit_price    7
dtype: int64

In [ ]:
filtered_product = df[df['sku'] == 'JBL0104']
num_unique_unit_prices = filtered_product['unit_price'].nunique()

num_unique_unit_prices

7

## Challenge 2:
List the (unique) items that were sold in the order with the id_order 385921.

In [ ]:
df.head()

id  id_order  product_id  product_quantity      sku unit_price  \
0  1119109    299539           0                 1  OTT0133      18.99   
1  1119110    299540           0                 1  LGE0043     399.00   
2  1119111    299541           0                 1  PAR0071     474.05   
3  1119112    299542           0                 1  WDT0315      68.39   
4  1119113    299543           0                 1  JBL0104      23.74   

                  date  
0  2017-01-01 00:07:19  
1  2017-01-01 00:19:45  
2  2017-01-01 00:20:57  
3  2017-01-01 00:51:40  
4  2017-01-01 01:06:38

In [ ]:
df.loc[df['id_order']== 385921, 'sku'].unique()

array(['APP2431', 'APP2348', 'APP2131', 'APP1630', 'APP1735', 'APP1216',
       'APP2092', 'APP1215', 'ELA0017', 'MIN0010', 'ELA0039', 'BEA0046',
       'BOS0034', 'BEA0071', 'ELA0029', 'APP2161', 'HOC0008', 'NOM0026',
       'NOM0014'], dtype=object)

In [ ]:
df.loc[df['id_order']== 385921,:]

id  id_order  product_id  product_quantity      sku unit_price  \
100355  1310787    385921           0                 1  APP2431   2.114.59   
100359  1310794    385921           0                 1  APP2348   3.805.59   
100360  1310796    385921           0                 1  APP2131     992.81   
100364  1310800    385921           0                 1  APP1630   1.020.33   
100365  1310801    385921           0                 1  APP1735     469.00   
100366  1310802    385921           0                 1  APP1216     139.99   
100368  1310804    385921           0                 1  APP2092     149.00   
100371  1310807    385921           0                 1  APP1215     110.99   
100374  1310810    385921           0                 1  ELA0017      32.99   
100379  1310815    385921           0                 2  MIN0010      69.99   
100384  1310821    385921           0                 1  ELA0039       9.99   
100386  1310823    385921           0                 1  BEA0046      99.90   
100387  1310825    385921           0                 1  BOS0034     249.99   
100390  1310828    385921           0                 1  BEA0071     269.99   
100391  1310830    385921           0                 2  ELA0029      29.99   
100392  1310832    385921           0                 1  APP2161     179.00   
100399  1310846    385921           0                 1  HOC0008      29.99   
100400  1310847    385921           0                 1  NOM0026      39.99   
100401  1310848    385921           0                 1  NOM0014      59.99   

                       date  
100355  2017-08-03 17:49:00  
100359  2017-08-03 17:54:39  
100360  2017-08-03 17:57:23  
100364  2017-08-03 18:04:06  
100365  2017-08-03 18:04:24  
100366  2017-08-03 18:05:04  
100368  2017-08-03 18:06:07  
100371  2017-08-03 18:07:18  
100374  2017-08-03 18:08:25  
100379  2017-08-03 18:10:24  
100384  2017-08-03 18:12:12  
100386  2017-08-03 18:12:59  
100387  2017-08-03 18:13:19  
100390  2017-08-03 18:14:24  
100391  2017-08-03 18:14:50  
100392  2017-08-03 18:17:26  
100399  2017-08-03 18:32:41  
100400  2017-08-03 18:33:21  
100401  2017-08-03 18:34:43

## Challenge 3:
Consider the products with the skus APP2431 and APP2348. Find out in how many orders they were present.

In [ ]:
df.loc[(df["sku"].isin(["APP2431", "APP2348"])), ["id_order"]].nunique()

id_order    179
dtype: int64

In [ ]:
df.loc[(df['sku'] == 'APP2431') | (df['sku'] == 'APP2348'), 'id_order'].nunique()

179

In [ ]:
df.loc[(df['sku']== 'APP2431') ,'id_order'].count() + df.loc[(df['sku']== 'APP2348') ,'id_order'].count()

189

In [ ]:
df.loc[(df['sku']== 'APP2431') & (df['sku'] == 'APP2348') ,'id_order']

Series([], Name: id_order, dtype: int64)

In [ ]:
(df.loc[df.sku.isin(['APP2431','APP2348'])]
                                  .groupby("id_order").agg({'sku':'count'})
                                  .sort_values('sku',ascending=False)
                                  .head(10)
                                  .index
                                  .to_list()
)

[379659,
 401302,
 382460,
 388422,
 402265,
 405858,
 410438,
 371564,
 367892,
 385921]

In [ ]:
special_orders = df.loc[df.sku.isin(['APP2431','APP2348'])].groupby("id_order").agg({'sku':'count'}).sort_values('sku',ascending=False).head(10).index.to_list()

In [ ]:
df.loc[df['id_order'].isin(special_orders) & df.sku.isin(['APP2431','APP2348']),:]

id  id_order  product_id  product_quantity      sku unit_price  \
79358   1272728    367892           0                 1  APP2431   2.108.00   
79359   1272730    367892           0                 1  APP2348   3.799.00   
83725   1280237    371564           0                 1  APP2348   3.799.00   
83729   1280243    371564           0                 1  APP2431   2.108.00   
93141   1297357    379659           0                 1  APP2348   3.719.00   
93144   1297362    379659           0                 1  APP2431   2.108.00   
96329   1303341    382460           0                 1  APP2431   2.114.59   
96337   1303357    382460           0                 1  APP2348   3.805.59   
100355  1310787    385921           0                 1  APP2431   2.114.59   
100359  1310794    385921           0                 1  APP2348   3.805.59   
103243  1315823    388422           0                 1  APP2431   2.114.59   
103244  1315824    388422           0                 1  APP2348   3.805.59   
118191  1354947    401302           0                 1  APP2431   2.114.59   
118193  1354949    401302           0                 1  APP2348   3.805.59   
119325  1357010    402265           0                 1  APP2431   2.114.59   
119326  1357011    402265           0                 1  APP2348   3.805.59   
123389  1364343    405858           0                 1  APP2348   3.805.59   
123390  1364344    405858           0                 1  APP2431   2.114.59   
128514  1373822    410438           0                 1  APP2431   2.114.59   
128515  1373825    410438           0                 1  APP2348   3.805.59   

                       date  
79358   2017-06-21 18:28:33  
79359   2017-06-21 18:30:37  
83725   2017-07-02 19:02:09  
83729   2017-07-02 19:07:35  
93141   2017-07-19 18:14:06  
93144   2017-07-19 18:14:48  
96329   2017-07-26 17:47:10  
96337   2017-07-26 18:00:06  
100355  2017-08-03 17:49:00  
100359  2017-08-03 17:54:39  
103243  2017-08-10 18:20:46  
103244  2017-08-10 18:21:37  
118191  2017-09-18 14:20:41  
118193  2017-09-18 14:21:21  
119325  2017-09-20 14:25:30  
119326  2017-09-20 14:25:53  
123389  2017-09-28 14:54:02  
123390  2017-09-28 14:54:19  
128514  2017-10-09 12:46:00  
128515  2017-10-09 12:48:30

## Challenge 4:
Create a new DataFrame, `df_50`, with all the rows that have a product quantity higher than 500 and only the columns `id`, `product_id`, `product_quantity` and `sku`. Be sure to use the method `.copy()`. Once the new DataFrame is created, modify the column "product_quantity" to "quantity", and "sku" to "product_code". To do so, you can use the method `.rename()` or assign a list of new names to the attribute `.columns`.

Creating `df_50`

In [ ]:
df_50 = df.loc[df['product_quantity']>=500,['id','product_id','product_quantity','sku']].copy()

Renaming columns of `df_50`

In [ ]:
df_50 = df_50.rename(columns={'product_quantity':'quantity','sku':'product_code'})

In [ ]:
df_50

id  product_id  quantity product_code
53860  1228150           0       999      APP1190
57306  1234111           0       555      APP0665
57796  1234924           0       800      KIN0137
68712  1254032           0       999      SEV0028

In [ ]:
df

id  id_order  product_id  product_quantity      sku unit_price  \
0       1119109    299539           0                 1  OTT0133      18.99   
1       1119110    299540           0                 1  LGE0043     399.00   
2       1119111    299541           0                 1  PAR0071     474.05   
3       1119112    299542           0                 1  WDT0315      68.39   
4       1119113    299543           0                 1  JBL0104      23.74   
...         ...       ...         ...               ...      ...        ...   
293978  1650199    527398           0                 1  JBL0122      42.99   
293979  1650200    527399           0                 1  PAC0653     141.58   
293980  1650201    527400           0                 2  APP0698       9.99   
293981  1650202    527388           0                 1  BEZ0204      19.99   
293982  1650203    527401           0                 1  APP0927      13.99   

                       date  
0       2017-01-01 00:07:19  
1       2017-01-01 00:19:45  
2       2017-01-01 00:20:57  
3       2017-01-01 00:51:40  
4       2017-01-01 01:06:38  
...                     ...  
293978  2018-03-14 13:57:25  
293979  2018-03-14 13:57:34  
293980  2018-03-14 13:57:41  
293981  2018-03-14 13:58:01  
293982  2018-03-14 13:58:36  

[293983 rows x 7 columns]

In [ ]:
new_func = lambda x : x+2

In [ ]:
new_func(4)

6

In [ ]:
#seid

df_50 = df.loc[lambda x: x['product_quantity'] > 500, ['id', 'product_id', 'product_quantity', 'sku'] ].copy()
df_50

id  product_id  product_quantity      sku
53860  1228150           0               999  APP1190
57306  1234111           0               555  APP0665
57796  1234924           0               800  KIN0137
68712  1254032           0               999  SEV0028

In [ ]:
df_50 = df[df["product_quantity"] > 500] [['id', 'product_id', 'product_quantity', 'sku']].copy()
df_50

id  product_id  product_quantity      sku
53860  1228150           0               999  APP1190
57306  1234111           0               555  APP0665
57796  1234924           0               800  KIN0137
68712  1254032           0               999  SEV0028

In [ ]:
df_50.columns = ['id', 'product_id', 'quantity', 'product_code']

In [ ]:
df_50

id  product_id  quantity product_code
53860  1228150           0       999      APP1190
57306  1234111           0       555      APP0665
57796  1234924           0       800      KIN0137
68712  1254032           0       999      SEV0028

In [ ]:
url = "https://drive.google.com/file/d/1afxwDXfl-7cQ_qLwyDitfcCx3u7WMvkU/view?usp=drive_link" # products.csv
path = "https://drive.google.com/uc?export=download&id="+url.split("/")[-2]
df_pr = pd.read_csv(path)

In [ ]:
df_pr['price'].isna().sum()

46

In [ ]:
df_pr.loc[~df_pr['price'].isna() & df_pr['name'].str.contains('case'),:]

sku                                               name  \
66     MAK0008  Maclocks theft case iPad 2 3 and 4 black with ...   
67     MAK0007  Maclocks theft case iPad 2 3 and 4 transparent...   
68     MAK0009     Maclocks theft case iPad 2 3 and 4 transparent   
69     MAK0010           Maclocks theft case iPad 2 3 and 4 black   
361    MUV0053               Muvit Slim case iPhone 5 / 5S purple   
...        ...                                                ...   
19020  ICA0111             Incase Frame Bumper Case iPhone X Gray   
19021  ICA0112        Incase Frame Bumper Case iPhone X Rose Gold   
19022  ICA0113           Incase Frame Bumper Case iPhone X Silver   
19031  ICA0114  Incase Travel Organizer Travel Organizer iPhon...   
19309  ICA0115       Incase City Brief case MacBook Pro 13 "Black   

                                                    desc  price promo_price  \
66     Case antitheft iPad 2 3 and 4 polycarbonate ro...    120   1.199.957   
67     Case antitheft iPad 2 3 and 4 polycarbonate ro...    120   1.079.961   
68     Case antitheft iPad 2 3 and 4 polycarbonate su...  79.99     799.931   
69     Case antitheft iPad 2 3 and 4 polycarbonate su...  79.99     799.931   
361     Imiatación Leather Case Cover for iPhone 5 / 5S.  14.99     119.911   
...                                                  ...    ...         ...   
19020  Transparent cover to protect the iPhone X with...  29.99     199.904   
19021  Transparent cover to protect the iPhone X with...  29.99     199.904   
19022  Transparent cover to protect the iPhone X with...  29.99     199.904   
19031  Sorts and secure your iPhone chargers and acce...  39.95     349.896   
19309      Incase bag to carry Apple MacBook Pro 13-inch  79.95     739.903   

       in_stock      type  
66            0  12635403  
67            0  12635403  
68            0  12635403  
69            0  12635403  
361           0  11865403  
...         ...       ...  
19020         1  11865403  
19021         1  11865403  
19022         1  11865403  
19031         1  11865403  
19309         1     10230  

[138 rows x 7 columns]

In [ ]:
df_pr.loc[df_pr['price'].isna()]

sku                                               name  \
34       TWS0019  Twelve South MagicWand support Apple Magic Tra...   
1900     AII0008           Aiino Case MacBook Air 11 '' Transparent   
2039     CEL0020  Celly Ambo Luxury Leather Case + iPhone 6 Case...   
2042     CEL0007  Celly Wallet Case with removable cover Black i...   
2043     CEL0012            Celly Silicone Hard Shell iPhone 6 Blue   
2044     CEL0014        Celly Silicone Hard Shell iPhone 6 Amarillo   
2049     CEL0015    Celly fur-lined Powerbank battery 4000mAh Black   
2051     CEL0018     Celly Wallet Leather Case cover Black iPhone 6   
2052     CEL0023  Celly Ambo Luxury Leather Case + Case Gold iPh...   
2053     CEL0025  Celly Ambo Luxury Leather Case + Case iPhone 6...   
8000     APP1197  Apple iMac 21.5 "Core i5 31 GHz Retina display...   
12052    CEL0028  Celly Swing Arm button selfie with Bluetooth B...   
12053    CEL0029  Celly extensible arm selfie with Bluetooth but...   
12054    CEL0030  Celly Smart Drive Housing + clip car holder iP...   
12055    CEL0031  Celly Smart Drive Case + Car clip support iPho...   
12059    CEL0032      Celly Gost clamp fixing car + Magnetica Black   
12060    CEL0042  Celly iPhone Case Fluor 6s / 6 Black / Transpa...   
12061    CEL0043  Celly Fluor Case iPhone 6 / 6S Pink / Transparent   
12062    CEL0044  Celly iPhone Case Fluor 6s / 6 Yellow / Transp...   
12064    CEL0038           Celly Frost Cover for iPhone 6 / 6S Rosa   
12067    CEL0033  Celly Waterproof iPad Case / Air / Pro 97 "Tra...   
12068    CEL0034  Celly Ambo cover with lid iPhone 6 / 6S Plus B...   
12069    CEL0035  Celly Ambo cover with lid iPhone 6 / 6S Plus B...   
12070    CEL0036  Celly Adapter Micro USB connector to Lightning...   
12071    CEL0051   Celly 4000mAh Battery Power Bank Aluminum Dorado   
12072    CEL0052   Celly 4000mAh Battery Power Bank Aluminum Silver   
12073    CEL0053             Celly IceCube Case iPhone 6 / 6S Black   
12074    CEL0054              Celly IceCube Case iPhone 6 / 6S Rosa   
12075    CEL0055             Celly IceCube Case iPhone 6 / 6S White   
12076    CEL0056          IceCube Celly iPhone Case 6 / 6s Amarillo   
12077    CEL0057        Celly IceCube Case iPhone 6 / 6S Plus Black   
12078    CEL0058         Celly IceCube Case iPhone 6 / 6S Plus Rosa   
12079    CEL0059        Celly IceCube Case iPhone 6 / 6S Plus White   
12080    CEL0060     Celly IceCube Case iPhone 6 / 6S Plus Amarillo   
12088    CEL0040     Celly Frost Cover for iPhone 6 / 6S Plus White   
12089    CEL0041              Celly iPhone 6S Plus Frost Case Black   
12092    CEL0045         Laser Case Celly iPhone 6 / 6S Dark Silver   
12093    CEL0046        Laser Case Celly iPhone 6 / 6S Clear Silver   
12095    CEL0047              Laser Case Celly iPhone 6 / 6S Dorado   
12096    CEL0048                Laser Case Celly iPhone 6 / 6S Rosa   
12097    CEL0049         Laser Case Celly iPhone 6 / 6S Plus Dorado   
12098    CEL0050   Laser Case Celly iPhone 6 / 6S Plus Silver Clear   
12100    CEL0037          Celly Frost Cover for iPhone 6 / 6S Black   
12238    CEL0027          Celly Frost Cover for iPhone 6 / 6S White   
12242    CEL0039       Celly Frost Cover for iPhone 6 / 6S Amarillo   
12947  CEL0034-A    (Open) Celly iPhone 6S Plus Housing Frost White   

                                                    desc price promo_price  \
34     MagicWand for wireless keyboard and Magic Trac...   NaN     299.899   
1900       MacBook Air 11-inch casing with matte finish.   NaN       22.99   
2039   Cover and housing together with magnet for iPh...   NaN     399.905   
2042              Case Book for iPhone 6 card case type.   NaN     128.998   
2043                       Hard Shell Silicone iPhone 6.   NaN        4.99   
2044                       Hard Shell Silicone iPhone 6.   NaN      59.895   
2049   Leather-wrapped External Battery 4000mAh for i...   NaN     239.895   
2051   Card case with transparent protective cover 

In [ ]:
df_pr.loc[df_pr['desc'].isna(),:]

sku                                               name desc  \
16126  WDT0211-A  Open - Purple 2TB WD 35 "PC Security Mac hard ...  NaN   
16128  APP1622-A  Open - Apple Smart Keyboard Pro Keyboard Folio...  NaN   
17843    PAC2334             Synology DS718 + NAS Server | 10GB RAM  NaN   
18152  KAN0034-A  Open - Kanex USB-C Gigabit Ethernet Adapter Ma...  NaN   
18490    HTE0025  Hyper Pearl 1600mAh battery Mini USB Mirror an...  NaN   
18612    OTT0200    OtterBox External Battery Power Pack 20000 mAHr  NaN   
18690  HOW0001-A  Open - Honeywell thermostat Lyric zonificador ...  NaN   

           price promo_price  in_stock      type  
16126        107     814.659         0      1298  
16128  1.568.206   1.568.206         0      1298  
17843     566.35   5.659.896         0  12175397  
18152      29.99     237.925         0      1298  
18490      24.99       22.99         1      1515  
18612      79.99       56.99         1      1515  
18690     199.99   1.441.174         0  11905404